Fix #192 ExpressReceiver support for rawBody for signature verification

seratch opened the pull request:
###  Summary

This pull request fixes issue #192 which prevents Bolt running on Google Cloud Platform.

I've extracted `verifySlackRequest` method from `ExpressReceiver` as `ExpressSignatureVerifier` and modified it a litle. Also, I changed the body parser `parseBody` in `ExpressReceiver` to handle `rawBody`. I'll add line comments to explain the intention further.

I haven't added unit tests for this change yet but I made sure the change works on real GCP environment. ~My example app on GCP managed to respond to `url_verification` requests.~
UPDATE: My example app on GCP responded to `url_verification` requests and handled events as I expect.

~Separately from that, it seems that we still have another issue after the body parsing step. My app on  GCP didn't work for event subscription (probably other requests, too) due to ack timeout errors. I've not figured out the root cause yet, but it seems to be hanging [here](https://github.com/slackapi/bolt/blob/%40slack/bolt%401.0.1/src/App.ts#L330-L332).~
^ UPDATE: Never mind. It was just my personal projects' quota issue...

### Requirements (place an `x` in each `[ ]`)

* [x] I've read and understood the [Contributing Guidelines](https://github.com/slackapi/bolt/blob/master/.github/contributing.md) and have done my best effort to follow them.
* [x] I've read and agree to the [Code of Conduct](https://slackhq.github.io/code-of-conduct).

seratch commented on src/ExpressReceiver.ts:
```diff
@@ -40,7 +39,7 @@ export default class ExpressReceiver extends EventEmitter implements Receiver {
 
     const expressMiddleware: RequestHandler[] = [
-      verifySlackRequest(signingSecret),
+      ExpressSignatureVerifier.create(signingSecret),
```
If you prefer other names, please let me know. I don't have any strong opinions on this.

seratch commented on src/ExpressReceiver.ts:
```diff
@@ -151,54 +150,39 @@ const respondToUrlVerification: RequestHandler = (req, res, next) => {
+  } else if (contentType === 'application/json') {
+    req.body = JSON.parse(stringBody);
   } else {
-    // TODO: should we check the content type header to make sure its JSON here?
```
Worked on this TODO but I'm willing to revert/modify this.

seratch commented on src/ExpressSignatureVerifier.ts:
```diff
@@ -0,0 +1,66 @@
+                        'Slack request signing verification failed. Timestamp is too old.',
+                        ErrorCode.ExpressReceiverAuthenticityError,
+                    );
+                    return next(error);
```
I realized that, even in this case, verifySlackRequest showed the following other errors. So, I added `return` here. If having `return` here has been intentionally avoided, let me know.

aoberoi replied:
Good catch! This must have been a bug.

seratch commented on src/ExpressSignatureVerifier.ts:
```diff
@@ -0,0 +1,66 @@
+
+                const signature = req.headers['x-slack-signature'] as string;
+                const ts = Number(req.headers['x-slack-request-timestamp']);
+                if (!signature && !ts) {
```
This is a newly added validation.

seratch commented on src/ExpressSignatureVerifier.ts:
```diff
@@ -0,0 +1,66 @@
+                        'Slack request signing verification failed. Signature mismatch.',
+                        ErrorCode.ExpressReceiverAuthenticityError,
+                    );
+                    return next(error);
```
same as above

aoberoi replied:
Again, a good catch! In the past, TypeScript has complained when I `return` the value returned by `next()`, and my solution was to just `return;` on its own on the next line. I'm surprised this is passing the type checker, but great!

seratch commented on src/ExpressReceiver.ts:
```diff
@@ -151,54 +150,39 @@ const respondToUrlVerification: RequestHandler = (req, res, next) => {
+  if (preparsedRawBody) {
+      stringBody = preparsedRawBody.toString();
+  } else {
+      stringBody = rawBody(req).toString();
```
A bug here. I'm writing tests and will fix this code, too.

seratch commented:
This pull request is now ready to be reviewed. I avoided doing force-push and pushed several commits. Please squash them into one when merging it.

aoberoi commented on src/ExpressReceiver.ts:
```diff
@@ -151,58 +150,52 @@ const respondToUrlVerification: RequestHandler = (req, res, next) => {
+ * Following middlewares can expect `req.body` is no long a string value.
+ */
+const parseBody: RequestHandler = async (req, _res, next) => {
+  const stringBody: string = await extractRequestBodyAsString(req);
```
why is this being called here instead of just accessing `req.body`? it seems that the `ExpressSignatureVerifier` will call this same function first, and assign a value back to `req.body`.

In fact, I think there are two concepts that we're trying to separate, but they are very coupled: request verification, and body parsing. Since they are so coupled, would it make sense to implement them in one `RequestHandler`? The one handler could call a helper function that knows how to verify the signature given a body `Buffer`, and a signing secret, a current time, and the values from the request header (signature and timestamp). Once the signature is verified, the `Buffer` contents can be trusted, so it's safe to parse the `Buffer` into an object and assign the result to `req.body`. I feel like the intermediate `string` representation isn't useful, and can be eliminated.

seratch replied:
Thank you for the clear explanation. Now I understand the point, and totally agree with the plan described here. I was also thinking that having temporary string representation is a bit confusing and actually has no benefit. I'm going to merge the two into one `RequestHandler` named `verifySignatureAndParseBody` (let me know if you have a better name for it).

aoberoi commented on src/ExpressReceiver.ts:
```diff
@@ -151,58 +150,52 @@ const respondToUrlVerification: RequestHandler = (req, res, next) => {
   } else {
-    // TODO: should we check the content type header to make sure its JSON here?
-    req.body = JSON.parse(req.body);
+    console.warn(`Unexpected content-type detected: ${contentType}`);
```
This is a great idea! Let's try not to use the `console` global from this file, and instead use a `Logger` instance. Since there isn't one available directly here, maybe this sort of concern should be the responsibility of the `ExpressReceiver` class (we can add a `logger` option to the constructor).

seratch replied:
Thanks for the suggestion. Will do.

aoberoi commented on src/ExpressReceiver.ts:
```diff
@@ -151,58 +150,52 @@ const respondToUrlVerification: RequestHandler = (req, res, next) => {
+      // Parse this body anyway
+      req.body = JSON.parse(stringBody);
+    } catch (e) {
+      console.error(`Failed to parse body as JSON data for content-type: ${contentType}`)
```
Same as above.

aoberoi commented:
@seratch This is really thorough and impressive. I have some comments and questions regarding the design, so I left them in this review. I did not look too closely at the tests because I think you might agree with some of the feedback, and those tests might chance along with the design. Happy to hear your thoughts.

seratch commented on src/ExpressReceiver.ts:
```diff
@@ -151,58 +150,52 @@ const respondToUrlVerification: RequestHandler = (req, res, next) => {
   next();
 };
 
-// TODO: this should be imported from another package
```
>If this package was distributed separately, (maybe @slack/verify-request), then people who don't use Bolt, or Express, or any other tool we provide could still use our request verification logic, and we would only have to maintain it once (instead of several places like it is today).

Sounds great. In this pull request, let me go with a necessary and sufficient patch for the issue happening on GCP. Afterwards, I'd love to be involved in that improvement, too.

aoberoi replied:
Great! I agree that we can come back on a separate pass to complete the extraction into a new package.

seratch commented on src/ExpressSignatureVerifier.ts:
```diff
@@ -0,0 +1,57 @@
+
+export default class ExpressSignatureVerifier {
+
+    public static create(signingSecret: string): RequestHandler {
```
Thank you for the comment. Yes, I have to admit this seems to be in a very other-lang style. 😅 

Anyway, as we discussed [here](https://github.com/slackapi/bolt/pull/197/files#r285785560), I'll merge this functionality into a single `RequestHandler` (which means merging this verification and parsing body process into one).

seratch commented:
@aoberoi Thank you very much for your generous review. I've updated the PR again.

Update summary:
* I merged `verifySlackRequest` and `parseBody` into unified `RequestHandler` named `verifySignatureAndParseBody`. One thing I should notice you is that I exported the function because I wanted to have tests for it. If you have suggestions or alternatives, please let me know.
* I added `logger` as an optional arg to `ExpressReceiverOptions` and `ExpressReceiver`'s constructor args + use it to output warnings/errors
* The function `verifyRequestSignature` is not ready to be extracted as a separate module but it's not exposed to external. Therefore, we can freely change the interface when extracting it as a reusable module.

aoberoi commented on src/ExpressReceiver.ts:
```diff
@@ -28,9 +30,10 @@ export default class ExpressReceiver extends EventEmitter implements Receiver {
+  constructor({
     signingSecret = '',
-    endpoints = { events: '/slack/events' },
+    logger = new ConsoleLogger(),
```
Can `App` pass in a `logger` when initializing this object? That way the `logLevel` is respected all the way down.

seratch replied:
Thank you for pointing out this. I should have been more careful about it... Anyway, I've updated the branch to fix this.

aoberoi commented:
This looks so good! Thanks especially for designing a testing solution for Express-related concerns, that looked tricky.
